data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java

```diff
@@ -75,43 +74,75 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
-                .orElseThrow(failure -> new InvalidRequestException(format("Request body was malformed: %s", failure.getFailureDetail())));
+        var transformResult = transformerRegistry.transform(expandedJson, CatalogRequestMessage.class);
+        if (transformResult.failed()) {
+            monitor.debug(String.format("%s, %s", BAD_REQUEST, verificationResult.getFailureMessages()));
+            return errorResponse(Response.Status.BAD_REQUEST, BAD_REQUEST);
+        }
 
+        var message = transformResult.getContent();
         // set protocol
         message.setProtocol(DATASPACE_PROTOCOL_HTTP);
 
-        var catalog = service.getCatalog(message, claimToken)
-                .orElseThrow(exceptionMapper(Catalog.class));
+        var claimToken = verificationResult.getContent();
 
-        var catalogJson = transformerRegistry.transform(catalog, JsonObject.class)
-                .orElseThrow(failure -> new EdcException(format("Failed to build response: %s", failure.getFailureDetail())));
+        var catalog = service.getCatalog(message, claimToken);
+        if (catalog.failed()) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP catalog error id %s: %s", errorCode, catalog.getFailureMessages()));
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
+        }
 
-        var compacted = jsonLdService.compact(catalogJson);
+        var catalogJson = transformerRegistry.transform(catalog.getContent(), JsonObject.class);
+        if (catalogJson.failed()) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP catalog error id %s: %s", errorCode, catalogJson.getFailureMessages()));
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
+        }
+
+        var compacted = jsonLdService.compact(catalogJson.getContent());
         if (compacted.failed()) {
-            throw new InvalidRequestException(compacted.getFailureDetail());
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP catalog error id %s: %s", errorCode, catalogJson.getFailureMessages()));
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
         }
-        //noinspection unchecked
-        return mapper.convertValue(compacted.getContent(), Map.class);
+
+        return Response.status(Response.Status.OK).type(MediaType.APPLICATION_JSON).entity(compacted.getContent()).build();
+    }
+
+    private Response errorResponse(Response.Status code, String message) {
+        return Response.status(code).type(MediaType.APPLICATION_JSON)
+                .entity(DspError.Builder.newInstance()
+                        .type(DSPACE_CATALOG_ERROR)
+                        .code(String.valueOf(code.getStatusCode()))
```

Comment:
Please no String.valueOf() :-)

Reply:
Lol, I was 100% sure I committed an update. Should be correct now.